libvast/vast/system/importer.hpp

```diff
@@ -29,27 +29,86 @@
 namespace vast::system {
 
 /// Receives chunks from SOURCEs, imbues them with an ID, and relays them to
-/// ARCHIVE and INDEX.
+/// ARCHIVE, INDEX and continuous queries.
 struct importer_state {
+  /// A simple generator for IDs.
+  struct id_generator {
+    /// The next available ID.
+    id i;
+
+    /// The first unavailable ID.
+    id last;
+
+    inline id_generator(id from, id to) : i(from), last(to) {
+      // nop
+    }
+
+    /// Returns whether this generator is exhausted.
+    inline bool at_end() const noexcept {
+      return i == last;
+    }
+
+    /// Returns the next ID and advances the position in the range.
+    inline id next() noexcept {
+      return i++;
+    }
+
+    /// Returns how many more times `next` returns a valid ID.
+    inline int32_t remaining() const noexcept {
+      return static_cast<int32_t>(last - i);
```

Comment:
Why `int32_t`? The semantics of this function don't seem to allow for negative values.